File: sdk-kotlin/src/main/kotlin/trade/decibel/sdk/utils/AddressUtils.kt
package trade.decibel.sdk.utils

import java.security.MessageDigest
import kotlin.math.ceil
import kotlin.math.floor
import kotlin.random.Random

object AddressUtils {

    fun getMarketAddr(name: String, perpEngineGlobalAddr: String): String {
        val addrBytes = hexToBytes(perpEngineGlobalAddr)
        val seed = bcsSerializeString(name)
        val objectAddr = createObjectAddress(addrBytes, seed)
        return "0x" + objectAddr.joinToString("") { "%02x".format(it) }
    }

    fun getPrimarySubaccountAddr(accountAddr: String, compatVersion: String, packageAddr: String): String {
        val addrBytes = hexToBytes(accountAddr)
        val seedStr = "${stripHexPrefix(packageAddr)}::dex_accounts::primary_account"
        val seed = seedStr.toByteArray()
        val objectAddr = createObjectAddress(addrBytes, seed)
        return "0x" + objectAddr.joinToString("") { "%02x".format(it) }
    }

    fun getVaultShareAddress(vaultAddress: String): String {
        val addrBytes = hexToBytes(vaultAddress)
        val seed = "vault_share".toByteArray()
        val objectAddr = createObjectAddress(addrBytes, seed)
        return "0x" + objectAddr.joinToString("") { "%02x".format(it) }
    }

    fun roundToTickSize(price: Double, tickSize: Double, pxDecimals: Int, roundUp: Boolean): Double {
        if (tickSize <= 0) return price
        val ticks = price / tickSize
        val roundedTicks = if (roundUp) ceil(ticks) else floor(ticks)
        return roundedTicks * tickSize
    }

    fun generateRandomReplayProtectionNonce(): ULong {
        return Random.nextLong().toULong()
    }

    private fun createObjectAddress(source: ByteArray, seed: ByteArray): ByteArray {
        val paddedSource = ByteArray(32)
        val srcLen = minOf(source.size, 32)
        System.arraycopy(source, 0, paddedSource, 32 - srcLen, srcLen)

        // Note: Using SHA-256 as placeholder. Aptos uses SHA3-256.
        val digest = MessageDigest.getInstance("SHA-256")
        digest.update(paddedSource)
        digest.update(seed)
        digest.update(byteArrayOf(0xFE.toByte()))
        return digest.digest()
    }

    private fun bcsSerializeString(s: String): ByteArray {
        val bytes = s.toByteArray()
        val result = mutableListOf<Byte>()
        var length = bytes.size
        do {
            var byte = (length and 0x7f).toByte()
            length = length shr 7
            if (length > 0) byte = (byte.toInt() or 0x80).toByte()
            result.add(byte)
        } while (length > 0)
        result.addAll(bytes.toList())
        return result.toByteArray()
    }

    private fun hexToBytes(hexStr: String): ByteArray {
        val stripped = stripHexPrefix(hexStr)
        val padded = if (stripped.length % 2 != 0) "0$stripped" else stripped
        return padded.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }

    private fun stripHexPrefix(s: String): String {
        return if (s.startsWith("0x")) s.substring(2) else s
    }
}

File: sdk-swift/Sources/DecibelSDK/Utils/AddressUtils.swift
import Foundation

public enum AddressUtils {
    /// Derives a market object address from the market name and perp engine global address.
    public static func getMarketAddr(name: String, perpEngineGlobalAddr: String) -> String {
        let addrBytes = hexToBytes(perpEngineGlobalAddr)
        let seed = bcsSerializeString(name)
        let objectAddr = createObjectAddress(source: addrBytes, seed: seed)
        return "0x" + objectAddr.map { String(format: "%02x", $0) }.joined()
    }

    /// Derives the primary subaccount address for an account.
    public static func getPrimarySubaccountAddr(accountAddr: String, compatVersion: String, packageAddr: String) -> String {
        let addrBytes = hexToBytes(accountAddr)
        let seedStr = "\(stripHexPrefix(packageAddr))::dex_accounts::primary_account"
        let seed = Array(seedStr.utf8)
        let objectAddr = createObjectAddress(source: addrBytes, seed: seed)
        return "0x" + objectAddr.map { String(format: "%02x", $0) }.joined()
    }

    /// Derives the vault share token address.
    public static func getVaultShareAddress(vaultAddress: String) -> String {
        let addrBytes = hexToBytes(vaultAddress)
        let seed = Array("vault_share".utf8)
        let objectAddr = createObjectAddress(source: addrBytes, seed: seed)
        return "0x" + objectAddr.map { String(format: "%02x", $0) }.joined()
    }

    /// Rounds a price to the nearest valid tick size.
    public static func roundToTickSize(price: Double, tickSize: Double, pxDecimals: Int32, roundUp: Bool) -> Double {
        guard tickSize > 0 else { return price }
        let ticks = price / tickSize
        let roundedTicks = roundUp ? ceil(ticks) : floor(ticks)
        return roundedTicks * tickSize
    }

    /// Generates a cryptographically secure random nonce for replay protection.
    public static func generateRandomReplayProtectionNonce() -> UInt64 {
        #if canImport(Security)
        var bytes = [UInt8](repeating: 0, count: 8)
        let status = SecRandomCopyBytes(kSecRandomDefault, 8, &bytes)
        if status == errSecSuccess {
            return bytes.withUnsafeBufferPointer {
                $0.baseAddress!.withMemoryRebound(to: UInt64.self, capacity: 1) { $0.pointee }
            }
        }
        // SecRandomCopyBytes failed; fall back to Swift.random
        return UInt64.random(in: UInt64.min...UInt64.max)
        #else
        return UInt64.random(in: UInt64.min...UInt64.max)
        #endif
    }

    // MARK: - Internal

    /// Aptos `create_object_address`: SHA3-256(source || seed || 0xFE)
    private static func createObjectAddress(source: [UInt8], seed: [UInt8]) -> [UInt8] {
        var paddedSource = [UInt8](repeating: 0, count: 32)
        let srcLen = min(source.count, 32)
        let startIndex = 32 - srcLen
        paddedSource[startIndex..<32] = source[0..<srcLen][...]

        var input = paddedSource
        input.append(contentsOf: seed)
        input.append(0xFE)
        return SHA3.sha256(input)
    }

    private static func bcsSerializeString(_ s: String) -> [UInt8] {
        let bytes = Array(s.utf8)
        var result: [UInt8] = []
        var length = bytes.count
        repeat {
            var byte = UInt8(length & 0x7f)
            length >>= 7
            if length > 0 { byte |= 0x80 }
            result.append(byte)
        } while length > 0
        result.append(contentsOf: bytes)
        return result
    }

    /// Converts a hex string (with optional 0x prefix) to bytes.
    /// Traps on invalid hex characters in debug; returns empty array in release.
    private static func hexToBytes(_ hexStr: String) -> [UInt8] {
        let stripped = stripHexPrefix(hexStr)
        let padded = stripped.count % 2 != 0 ? "0" + stripped : stripped
        var bytes: [UInt8] = []
        bytes.reserveCapacity(padded.count / 2)
        var index = padded.startIndex
        while index < padded.endIndex {
            let nextIndex = padded.index(index, offsetBy: 2)
            guard let byte = UInt8(padded[index..<nextIndex], radix: 16) else {
                assertionFailure("Invalid hex byte in \"\(hexStr)\" at offset \(padded.distance(from: padded.startIndex, to: index))")
                return bytes
            }
            bytes.append(byte)
            index = nextIndex
        }
        return bytes
    }

    private static func stripHexPrefix(_ s: String) -> String {
        s.hasPrefix("0x") ? String(s.dropFirst(2)) : s
    }
}
